File: web/app/(doctor)/doctor/_lib/dictation.ts
// Typed client for the dictation surface (backend app/routes/dictation.py).
//
// Mirrors the backend's deliberate split into verbs: `start` stores the
// transcript, `mapFields` runs the model, `correct` records the doctor's fixes,
// `sign` locks it. The split is what lets a failed mapping keep the recording,
// so the client keeps it too rather than collapsing them into one save.

import { API_BASE, AuthError } from "@/app/_lib/queue";

export type Suggestion = { name: string; generic: string; score: number };

export type Med = {
  /** Exactly what the doctor said. The UI must never substitute a formulary name here. */
  name: string;
  dose: string | null;
  route: string | null;
  freq: string | null;
  duration: string | null;
  as_spoken: string;
  known: boolean;
  generic: string | null;
  drug_class: string | null;
  ambiguous: boolean;
  /** Fuzzy neighbours from the formulary. Advice for the doctor — never a value. */
  suggestions: Suggestion[];
  /** The name is not in the transcript — the model renamed or invented it. */
  unsaid: boolean;
  acknowledged: boolean;
};

export type TreatmentEvent = {
  cycle: number | null;
  regimen: string;
  date: string | null;
  next_due: string | null;
  as_spoken: string;
};

export type FollowUp = { when: string | null; as_spoken: string; instructions: string };

/** The doctor's own ayurvedic assessment (doc 24 §6.1).
 *
 *  Five free-text lines, and every one of them is typed by the doctor: no model
 *  writes these, the prompts are told never to infer a prakriti or a dosha, and
 *  there is no server path that fills them from a mapping. Optional on the wire
 *  so a note stored before doc 24 parses. */
export type Assessment = {
  prakriti: string;
  vikriti: string;
  agni: string;
  koshtha: string;
  nidana: string;
};

export type MappedFields = {
  diagnosis: string | null;
  treatment_events: TreatmentEvent[];
  meds: Med[];
  advice: string[];
  follow_up: FollowUp;
  unclear: string[];
  assessment?: Assessment;
  /** Diet & lifestyle lines (doc 24 §6.2). Kept apart from `advice` because they
   *  print under their own heading, and in an ayurveda consult they are half of
   *  the treatment rather than a footnote to it. */
  pathya_apathya?: string[];
};

/** An assessment with every line blank — what a note that has none looks like,
 *  and the starting point for one a doctor is about to fill in. */
export const EMPTY_ASSESSMENT: Assessment = {
  prakriti: "",
  vikriti: "",
  agni: "",
  koshtha: "",
  nidana: "",
};

export type Dictation = {
  id: string;
  visit_id: string;
  status: "draft" | "signed";
  transcript: string | null;
  /** What the model produced — frozen, so the review can diff against it. */
  mapped: MappedFields | null;
  /** What the record says now, after corrections. */
  fields: MappedFields | null;
  edits: { at: string; by: string; field: string }[];
  model: string | null;
  prompt_ref: string | null;
  mapping_error: string | null;
  mapped_at: string | null;
  signed_at: string | null;
  /** Flagged drugs still blocking the signature. */
  blocking_meds: string[];
};

function authHeaders(token: string): HeadersInit {
  return { Authorization: `Bearer ${token}`, "Content-Type": "application/json" };
}

async function unwrap(res: Response): Promise<Dictation> {
  if (res.status === 401) throw new AuthError();
  if (!res.ok) {
    let detail = `dictation ${res.status}`;
    try {
      const body = await res.json();
      if (typeof body?.detail === "string") detail = body.detail;
    } catch {
      /* keep the status-code message */
    }
    throw new Error(detail);
  }
  return res.json();
}

export async function fetchDictation(token: string, visitId: string): Promise<Dictation | null> {
  const res = await fetch(`${API_BASE}/dictation/visits/${visitId}`, {
    headers: authHeaders(token),
    cache: "no-store",
  });
  if (res.status === 401) throw new AuthError();
  if (!res.ok) throw new Error(`dictation ${res.status}`);
  return res.json();
}

export async function startDictation(
  token: string,
  visitId: string,
  transcript: string,
): Promise<Dictation> {
  return unwrap(
    await fetch(`${API_BASE}/dictation/visits/${visitId}`, {
      method: "POST",
      headers: authHeaders(token),
      body: JSON.stringify({ transcript }),
    }),
  );
}

/**
 * "Type note": open the editable fields with no model in the loop.
 *
 * The same record, the same corrections trail, the same signature and the same
 * refusal on a drug the formulary does not know. Speech is an input method here,
 * not a prerequisite for prescribing — which is what it was until now.
 */
export async function composeNote(token: string, dictationId: string): Promise<Dictation> {
  return unwrap(
    await fetch(`${API_BASE}/dictation/${dictationId}/compose`, {
      method: "POST",
      headers: authHeaders(token),
    }),
  );
}

export async function mapFields(token: string, dictationId: string): Promise<Dictation> {
  return unwrap(
    await fetch(`${API_BASE}/dictation/${dictationId}/map`, {
      method: "POST",
      headers: authHeaders(token),
    }),
  );
}

/** What "tap to fix" may send.
 *
 *  `Partial<MappedFields>` everywhere except `assessment`, which is a
 *  **partial of a partial**: the five lines are edited one at a time and the
 *  server merges them by key, so a commit carries only the line that changed.
 *  Sending the whole object would spread the other four from state one round
 *  trip old and blank whichever the doctor filled in just before this one. */
export type NotePatch = Omit<Partial<MappedFields>, "assessment"> & {
  assessment?: Partial<Assessment>;
};

export async function correct(
  token: string,
  dictationId: string,
  patch: NotePatch,
): Promise<Dictation> {
  return unwrap(
    await fetch(`${API_BASE}/dictation/${dictationId}`, {
      method: "PATCH",
      headers: authHeaders(token),
      body: JSON.stringify(patch),
    }),
  );
}

export async function signDictation(token: string, dictationId: string): Promise<Dictation> {
  return unwrap(
    await fetch(`${API_BASE}/dictation/${dictationId}/sign`, {
      method: "POST",
      headers: authHeaders(token),
    }),
  );
}

// The STT call moved to `useVoiceCapture` in M4, where the recorder that makes
// the recording now lives. Both doctor surfaces post to it; they differ only in
// which endpoint, and therefore which meter, receives the clip.

File: backend/app/routes/dictation.py
"""The dictation surface (doc 03 §7).

Every route is `require_doctor` — tighter than the coordinator's
`require_staff`, for the same reason the S9 card is: this is a patient's
diagnosis and their prescription, and a queue coordinator has no business in it.
And unlike `/kiosk/stt`, `/dictation/stt` is authenticated: a kiosk clip is an
anonymous chief complaint from a public terminal, while this clip is a named
patient's consult.

The verbs are deliberately separate rather than one save-everything endpoint:

    POST   /dictation/visits/{visit_id}      open the draft, store the transcript
    POST   /dictation/{id}/map               transcript -> structured fields
    PATCH  /dictation/{id}                   the doctor's corrections
    POST   /dictation/{id}/sign              lock it
    GET    /dictation/visits/{visit_id}      read it back
    POST   /dictation/stt                    audio -> transcript (Web Speech fallback)

Splitting map from save is what lets a failed mapping keep the transcript: the
recording is the part that cannot be recreated once the doctor has moved to the
next patient.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal, InvalidOperation
from typing import Any

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app import dictation as dictation_svc
from app import doctor as doctor_svc
from app.auth.rbac import Principal, require_doctor
from app.config import Settings, get_settings
from app.db import get_session
from app.models.clinical import Dictation
from app.models.enums import Lang, UsagePurpose
from app.providers import AudioClip, ProviderBadRequest, ProviderError, with_fallback
from app.providers.metering import usage_scope
from app.providers.registry import llm_chain, stt_chain

router = APIRouter(prefix="/dictation", tags=["dictation"])

#: A consult note is a minute or two of speech, not a lecture. Generous enough
#: for a long oncology plan, small enough that a stuck recorder cannot post a
#: gigabyte at the box's Whisper.
_MAX_STT_BYTES = 24 * 1024 * 1024


# -- wire models --------------------------------------------------------------


class SuggestionOut(BaseModel):
    name: str
    generic: str
    score: float


class MedOut(BaseModel):
    #: Exactly what the doctor said. Never a formulary name (doc 03 §7).
    name: str
    dose: str | None = None
    route: str | None = None
    freq: str | None = None
    duration: str | None = None
    as_spoken: str = ""
    known: bool = False
    generic: str | None = None
    drug_class: str | None = None
    ambiguous: bool = False
    suggestions: list[SuggestionOut] = []
    #: The name is not in the transcript — the model renamed or invented it.
    unsaid: bool = False
    acknowledged: bool = False


class TreatmentEventOut(BaseModel):
    cycle: int | None = None
    regimen: str = ""
    date: str | None = None
    next_due: str | None = None
    as_spoken: str = ""


class FollowUpOut(BaseModel):
    when: str | None = None
    as_spoken: str = ""
    instructions: str = ""


class MappingOut(BaseModel):
    diagnosis: str | None = None
    treatment_events: list[TreatmentEventOut] = []
    meds: list[MedOut] = []
    advice: list[str] = []
    follow_up: FollowUpOut = FollowUpOut()
    unclear: list[str] = []


class EditOut(BaseModel):
    at: str
    by: str
    field: str


class DictationOut(BaseModel):
    id: uuid.UUID
    visit_id: uuid.UUID
    status: str
    transcript: str | None = None
    #: What the model produced. Frozen — the review screen diffs against it.
    mapped: MappingOut | None = None
    #: What the record says now, after the doctor's corrections.
    fields: MappingOut | None = None
    edits: list[EditOut] = []
    model: str | None = None
    prompt_ref: str | None = None
    mapping_error: str | None = None
    mapped_at: str | None = None
    signed_at: datetime | None = None
    #: Flagged drugs the doctor has not yet acknowledged — unrecognised, or
    #: named something they are not recorded as having said. The exact list
    #: `sign` will refuse on.
    blocking_meds: list[str] = []


class StartIn(BaseModel):
    transcript: str = Field(default="", max_length=20_000)


class PatchIn(BaseModel):
    """Whole-field replacement; only the doc 03 §7 fields are editable."""

    diagnosis: str | None = None
    treatment_events: list[dict[str, Any]] | None = None
    meds: list[dict[str, Any]] | None = None
    advice: list[str] | None = None
    follow_up: dict[str, Any] | None = None
    unclear: list[str] | None = None

    def patch(self) -> dict[str, Any]:
        return self.model_dump(exclude_unset=True)


class SttOut(BaseModel):
    text: str
    provider: str
    lang: str
    confidence: float | None = None
    uncertain: bool = False


# -- serialisation ------------------------------------------------------------


def _out(dictation: Dictation) -> DictationOut:
    structured = dictation.structured or {}
    mapping = dictation_svc.current_mapping(dictation)
    blocking = [m.name for m in mapping.meds_needing_attention] if mapping else []
    return DictationOut(
        id=dictation.id,
        visit_id=dictation.visit_id,
        status=str(dictation.status),
        transcript=dictation.transcript,
        mapped=structured.get("mapped"),
        fields=structured.get("fields"),
        edits=[
            EditOut(at=str(e.get("at")), by=str(e.get("by")), field=str(e.get("field")))
            for e in structured.get("edits") or []
        ],
        model=structured.get("model"),
        prompt_ref=structured.get("prompt_ref"),
        mapping_error=structured.get("mapping_error"),
        mapped_at=structured.get("mapped_at"),
        signed_at=dictation.signed_at,
        blocking_meds=blocking,
    )


async def _doctor(session: AsyncSession, principal: Principal):
    try:
        return await doctor_svc.resolve_doctor(session, user_id=principal.id)
    except doctor_svc.DoctorError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc


async def _load(session: AsyncSession, dictation_id: uuid.UUID, doctor) -> Dictation:
    dictation = await session.get(Dictation, dictation_id)
    if dictation is None or dictation.deleted_at is not None:
        raise HTTPException(status_code=404, detail="no such dictation")
    # Scope by the visit, not by `doctor_id`: a colleague covering the room reads
    # the same department's notes, and the S9 card already draws that boundary.
    try:
        await dictation_svc.assert_visit_scope(session, visit_id=dictation.visit_id, doctor=doctor)
    except dictation_svc.DictationError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc
    return dictation


def _fail(exc: dictation_svc.DictationError) -> HTTPException:
    if isinstance(exc, dictation_svc.DictationLocked):
        return HTTPException(status_code=409, detail=str(exc))
    if isinstance(exc, dictation_svc.MappingUnavailable):
        return HTTPException(status_code=503, detail=str(exc))
    return HTTPException(status_code=400, detail=str(exc))


# -- routes -------------------------------------------------------------------


@router.get("/visits/{visit_id}", response_model=DictationOut | None)
async def read(
    visit_id: uuid.UUID,
    principal: Principal = Depends(require_doctor),
    session: AsyncSession = Depends(get_session),
) -> DictationOut | None:
    """This visit's consult note, or null if none has been started."""
    doctor = await _doctor(session, principal)
    try:
        dictation = await dictation_svc.get_draft(session, visit_id=visit_id, doctor=doctor)
    except dictation_svc.DictationError as exc:
        raise _fail(exc) from exc
    return _out(dictation) if dictation else None


@router.post("/visits/{visit_id}", response_model=DictationOut)
async def start(
    visit_id: uuid.UUID,
    body: StartIn,
    principal: Principal = Depends(require_doctor),
    session: AsyncSession = Depends(get_session),
) -> DictationOut:
    """Open the draft and store the transcript. Idempotent per visit."""
    doctor = await _doctor(session, principal)
    try:
        dictation = await dictation_svc.start(
            session, visit_id=visit_id, doctor=doctor, transcript=body.transcript
        )
    except dictation_svc.DictationError as exc:
        raise _fail(exc) from exc
    return _out(dictation)


@router.post("/{dictation_id}/map", response_model=DictationOut)
async def map_fields(
    dictation_id: uuid.UUID,
    principal: Principal = Depends(require_doctor),
    session: AsyncSession = Depends(get_session),
    settings: Settings = Depends(get_settings),
) -> DictationOut:
    """Transcript → structured fields, on whichever LLM is configured.

    Cloud Gemini Flash or the box's own Qwen3 (`LLM_PROVIDER=local_vllm`) — same
    code, same prompt, same formulary validation afterwards.
    """
    doctor = await _doctor(session, principal)
    dictation = await _load(session, dictation_id, doctor)
    mapper = dictation_svc.DictationMapper(llm_chain(settings))
    try:
        # Attributed to the visit, not to a channel: dictation is not a patient
        # channel, and the S18 dashboard wants this rupee amount next to the
        # intake it belongs to.
        with usage_scope(visit_id=dictation.visit_id):
            dictation = await dictation_svc.map_transcript(
                session, dictation=dictation, doctor=doctor, mapper=mapper
            )
    except dictation_svc.DictationError as exc:
        raise _fail(exc) from exc
    return _out(dictation)


@router.patch("/{dictation_id}", response_model=DictationOut)
async def correct(
    dictation_id: uuid.UUID,
    body: PatchIn,
    principal: Principal = Depends(require_doctor),
    session: AsyncSession = Depends(get_session),
) -> DictationOut:
    """The doctor's tap-to-fix. Re-validates any drug name they typed."""
    doctor = await _doctor(session, principal)
    dictation = await _load(session, dictation_id, doctor)
    try:
        dictation = await dictation_svc.apply_corrections(
            session, dictation=dictation, doctor=doctor, patch=body.patch()
        )
    except dictation_svc.DictationError as exc:
        raise _fail(exc) from exc
    return _out(dictation)


@router.post("/{dictation_id}/sign", response_model=DictationOut)
async def sign(
    dictation_id: uuid.UUID,
    principal: Principal = Depends(require_doctor),
    session: AsyncSession = Depends(get_session),
) -> DictationOut:
    """Sign and lock. 400 while an unrecognised drug is unacknowledged."""
    doctor = await _doctor(session, principal)
    dictation = await _load(session, dictation_id, doctor)
    try:
        dictation = await dictation_svc.sign(session, dictation=dictation, doctor=doctor)
    except dictation_svc.DictationError as exc:
        raise _fail(exc) from exc
    return _out(dictation)


@router.post("/stt", response_model=SttOut)
async def stt(
    file: UploadFile = File(...),
    lang: Lang = Form(Lang.EN),
    duration_seconds: str | None = Form(default=None),
    principal: Principal = Depends(require_doctor),
    settings: Settings = Depends(get_settings),
) -> SttOut:
    """The accuracy pass behind Web Speech (doc 03 §7).

    Chrome's Web Speech API is the fast path in the console, but it ships the
    doctor's voice to a cloud recogniser and it is poor at Hinglish drug names.
    This route is the fallback and, on a V-OSS box, the better one: the
    configured chain is local Whisper, so the consult never leaves the premises.
    """
    data = await file.read()
    if not data:
        raise HTTPException(status_code=422, detail="empty audio upload")
    if len(data) > _MAX_STT_BYTES:
        raise HTTPException(status_code=413, detail="recording too large")

    duration: Decimal | None = None
    if duration_seconds:
        try:
            duration = Decimal(duration_seconds)
        except (InvalidOperation, ValueError):
            duration = None

    clip = AudioClip(data=data, mime=file.content_type or "audio/webm", duration_seconds=duration)
    try:
        with usage_scope():
            transcript = await with_fallback(
                stt_chain(settings),
                lambda p: p.transcribe(clip, str(lang), purpose=UsagePurpose.DICTATION),
            )
    except ProviderBadRequest as exc:
        raise HTTPException(status_code=422, detail=f"could not read that audio: {exc}") from exc
    except ProviderError as exc:
        raise HTTPException(status_code=503, detail="speech recognition is unavailable") from exc

    return SttOut(
        text=transcript.text,
        provider=transcript.provider,
        lang=transcript.lang,
        confidence=transcript.confidence,
        uncertain=transcript.is_uncertain,
    )
